### PR TITLE
bzl: rework how we populate the database schemas for migrator

### PR DESCRIPTION
**Pull request by jhchabran**
In the context of RFC 795, schemas are a perpetual source of trouble as we currently deal with them. I tried a few approaches, but there is always an edge case. 

This PR introduces a different approach where we offload all the work to a GCS bucket at build time (previously we fetched things from GitHub through HTTP requests to their CDN).  

Please see `tools/release/README.md` for a complete description, including a diagram and a step by step approach. 

Due to the critical nature of the tarballs in our release process, they are built in a way that make them idempotent, meaning that if you run the script three times, the resulting tarball will always have the same checksum. That makes it possible to regenerate tarballs for previous versions if something were to go bad with the bucket. 

As for the problem that @unknwon pointed out, i.e. running two release builds at the same time, like 5.1.999 and 5.2.333 you could end up with 5.2.333 not being aware of the existence of 5.1.999, this is not addressed in that PR because this is a release tooling issue, which we can address later. 

To sum up how this works both now and tomorrow: 

- Now 
  - The database schemas are up to date and take into account the recent 5.2.1 release. 
  - If we are cutting a new 5.2.2 release, for example, nothing needs to be done on the 5.2.x branch, the only thing needed is for *me* to upload that new database schemas in the bucket manually. 
  - We can safely use this on `main`, this changes won't reach customers before 5.3.X. 
- After
  - Once I've rolled out the new release process, this script will be called automatically by the release tooling. 
  - This is where the concern from @unknwon will be covered. 

## Test plan

Locally tested, ran the `//cmd/migrator:image_test` and they passed.

### CODE REVIEW

File: tools/release/generate_schemas_archive.sh [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+echo "--- Creating tarball '$output'"
+tar cvzf "$output" -C "$tmp_dir" $(ls -A $tmp_dir)
+checksum=$(sha256sum "$output")
+echo "Checksum: $checksum"
+
+echo "--- Uploading tarball to ${bucket}/dist"
+gsutil cp "$output" "${bucket}/dist/"
+
+echo "--- Updating buildfiles"
+
```
Comment (jhchabran) [inline tools/release/generate_schemas_archive.sh:74]: TODO

File: tools/release/README.md [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+1. t=0 5.1.0 has been released publicly
+  - `main` branch is now considered to be 5.2.0
+  - `5.1` branch is the target for PRs for backports and bug fixes.
+1. t=10 5.1.2222 has been released publicly 
+  - `5.1` branch is from where this release was cut.
+2. t=20 5.2.0 has been released publicly
+  - `main` branch is now considered to be 5.3.0
+  - `5.2` branch is the target for PRs for backports and bug fixes.
+3. t=30 5.1.3333 has been released publicly
+  - `5.2` branch is from where this release was cut.
```
Comment (unknwon) [inline tools/release/README.md:70]:
```suggestion
1. t=0 5.1.0 has been released publicly
   - `main` branch is now considered to be 5.2.0
   - `5.1` branch is the target for PRs for backports and bug fixes.
1. t=10 5.1.2222 has been released publicly 
   - `5.1` branch is from where this release was cut.
2. t=20 5.2.0 has been released publicly
   - `main` branch is now considered to be 5.3.0
   - `5.2` branch is the target for PRs for backports and bug fixes.
3. t=30 5.1.3333 has been released publicly
   - `5.2` branch is from where this release was cut.
```

Don't ask me why 3 spaces, I just found 2 is not working from my experience.

Comment (jhchabran) [reply tools/release/README.md:70]: What also infuriates me is that you need four spaces to get sub-bullet points on Slack.

File: tools/release/README.md [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+ │  │ dist/                             │  │
+ │  │  schemas-v5.2.1093.tar.gz         │  │
+ │  │  schemas-v5.2.1234.tar.gz ◄───────┼──┼──────  Uploaded at the beginning of
+ │  │             ▲                     │  │        of a new release build.
+ │  │             │                     │  │
+ │  │             │                     │  │        Release build automatically
+ │  │             │                     │  │        update the Bazel reference
+ │  │             │                     │  │        to that file.
+ │  └─────────────┼─────────────────────┘  │
+ │                │                        │        Gets deleted if the release
```
Comment (unknwon) [inline tools/release/README.md:34]:
```suggestion
 │  │             ▲                     │  │        of a new normal build.
 │  │             │                     │  │
 │  │             │                     │  │        Release build automatically
 │  │             │                     │  │        update the Bazel reference
 │  │             │                     │  │        to that file.
 │  └─────────────┼─────────────────────┘  │
 │                │                        │        Gets deleted if the normal
```

From my understanding of the text in the rest of the document, here should be "normal build" not "release build"? 🤔

File: tools/release/README.md [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+
+1. t=0 5.1.0 has been released publicly
+  - `main` branch is now considered to be 5.2.0
+  - `5.1` branch is the target for PRs for backports and bug fixes.
+1. t=10 5.1.2222 has been released publicly 
+  - `5.1` branch is from where this release was cut.
+2. t=20 5.2.0 has been released publicly
+  - `main` branch is now considered to be 5.3.0
+  - `5.2` branch is the target for PRs for backports and bug fixes.
+3. t=30 5.1.3333 has been released publicly
```
Comment (unknwon) [inline tools/release/README.md:69]:
```suggestion
3. t=30 5.2.3333 has been released publicly
```

You mean `5.2.x`?

File: tools/release/README.md [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+  - the updated buildfiles will now use that uploaded `schemas-v5.3.4444.tar.gz` tarball from now on, eliminating the need to fetch anything from GCS apart the tarball (until it's cached by Bazel).
+
+## Q&A 
+
+> What happens if two release builds are built at the same time?
+
+If two builds are going on at the same time, they won't interfere with each other, because the only artifacts that can be removed without notice are the schemas tarballs, which are 
+only referenced by each individual release build. As for the schemas, the only time they get created is when the internal release is finally green and ready to be merged. If one of the two builds end 
+up referencing the schemas from the other, it means they didn't happen at the same time, but instead that they happened sequentially. That's because GCS is guaranteeing us that file uploads are 
+transactional, i.e it's not possible to list a file until it's fully uploaded.
```
Comment (unknwon) [inline tools/release/README.md:114]: > That's because GCS is guaranteeing us that file uploads are 
transactional, i.e it's not possible to list a file until it's fully uploaded.

Hmm, isn't this actually the _source_ of the problem instead of the _reason_ that's not a problem?

If say, 5.1.1111 and 5.2.2222 kicks off at the same time, and assuming in the end they both green and will be marked as public releases. By definition, 5.2.2222 need to include schemas change from 5.1.1111, but 5.2.2222 runs faster (for reasons) and missed it while 5.1.1111 is still uploading its schemas.

File: tools/release/README.md [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+  - `5.1` branch is the target for PRs for backports and bug fixes.
+1. t=10 5.1.2222 has been released publicly 
+  - `5.1` branch is from where this release was cut.
+2. t=20 5.2.0 has been released publicly
+  - `main` branch is now considered to be 5.3.0
+  - `5.2` branch is the target for PRs for backports and bug fixes.
+3. t=30 5.1.3333 has been released publicly
+  - `5.2` branch is from where this release was cut.
+  
+So with that scenario, when 5.1.3333 has been released, we introduced a new version that the _migrator_ must be aware of, on both `main` and the `5.1` branch. Previously, this required to make a PR to port from the 5.1 branch the references
```
Comment (DaedalusG) [inline tools/release/README.md:72]:
```suggestion
So with that scenario, when 5.1.3333 has been released, we introduced a new version that the _migrator_ must be aware of, on both `main` and the `5.1` branch. Previously, this required us to make a PR to port to main, the 5.1 branch references 
```

File: tools/release/README.md [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+
+If two builds are going on at the same time, they won't interfere with each other, because the only artifacts that can be removed without notice are the schemas tarballs, which are 
+only referenced by each individual release build. As for the schemas, the only time they get created is when the internal release is finally green and ready to be merged. If one of the two builds end 
+up referencing the schemas from the other, it means they didn't happen at the same time, but instead that they happened sequentially. That's because GCS is guaranteeing us that file uploads are 
+transactional, i.e it's not possible to list a file until it's fully uploaded. 
+
+> What happens if a release build fails. Can it mess with ulterior release builds?
+
+It cannot, because the only time the schemas are finally added to `schemas/` is when the release build succeeds. This is why when we're regenerating the tarball, we are fetching 
+all the known schemas *and* adding the new one from the source tree at that point. Had we had uploaded the new schemas at the beginning of the build instead, to then fetch everything to
```
Comment (DaedalusG) [inline tools/release/README.md:119]:
```suggestion
all the known schemas *and* adding the new one from the source tree at that point. Had we uploaded the new schemas at the beginning of the build instead, to then fetch everything to  
```

File: tools/release/README.md [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+
+> How do we ensure that the `schema.*.json` in the source, at the revision we're cutting the release are correct? 
+
+This is covered by Bazel. These files are generated through `bazel run //dev:write_all_generated` which comes with automatically generated `diff_test` rules, which are comparing 
+the files on disk, with the files it would generate. Therefore, if someone pushes code without updating the current schemas in the code, Bazel will fail the build. And if on that
+precise commit we would try to cut a release, that same exact test would run again and fail.
+
+Therefore, we can safely use the current schemas when cutting a release. 
+
+> What happens if the _migrator_ is built with newer schemas, like 5.1.3333 that contains schemas for 5.2.4444?
```
Comment (DaedalusG) [inline tools/release/README.md:130]: Why would this happen anyway?

File: tools/release/README.md [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+- After
+  - Cutting a new release
+    - Schemas are always up to date when cutting a new release. No need to port changes to other release branches or `main`.
+    - Schemas are downloaded concurrently - only takes a few second to grab all of them.
+  - When building normally
+    - Schemas are cached if the Bazel cache is warm. Otherwise, we download a single tarball of a few mbs.
+
+> How do I see which schemas where used to build the _migrator_ container.
+
+`tar tf $(bazel cquery //cmd/migrator:tar_schema_descriptions --output=files)` will show the content the container layer used
```
Comment (DaedalusG) [inline tools/release/README.md:157]: Nice

**Comment by DaedalusG** [comment]
This seems like it will work great to me. One thing to note is that migrator is using the gcs bucket schemas as a fallback and we [prefer](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/internal/database/migration/schemas/drift_schema.go?L16-22) the schemas to be embedded in the build. This should work the same though.

After a build is marked green we can run an upgrade smoke test before marking the build as good for release. 👍🏼

File: tools/release/README.md (hunk 8 above)
Comment (jhchabran) [reply tools/release/README.md:130]: Because the code populating the schemas would blindly fetch all available schemas instead of filtering them out as I did in the code from this PR. I'm mentioning it because it's tempting to assume the migrator would exclude the schemas it doesn't need (I'm not sure of that, I wanted to ask you first :) )

File: tools/release/README.md (hunk 3 above)
Comment (jhchabran) [reply tools/release/README.md:34]: Normal build is a random CI build on `main`. You never upload anything in that scenario.

File: tools/release/README.md [hunk 10]
```diff
@@ -0,0 +1,158 @@
+# Release tooling
```
Comment (keegancsmith) [inline tools/release/README.md:1]: this whole README is beautiful

File: tools/release/README.md [hunk 11]
```diff
@@ -0,0 +1,198 @@
...
+1. it updates `tools/release/schema_deps.bzl` with the new tarball URL and its checksum.
+1. CI builds the new release.
+1. At the end of the build:
+
+- If green
+  - the schemas `v5.3.4444-internal_database.schema.*.json` are uploaded to the `schemas/` folder.
+- If red
+  - the schemas `v5.3.4444-internal_database.schema.*.json` are _NOT_ uploaded to the `schemas/` folder.
+    - that's because if the release build failed, it means that it never existed, so there is no need to capture its existence as nobody will migrate from that version number.
+  - the `schemas-v5.3.4444.tar.gz` tarball is removed from the `dist/` folder in the bucket. This is perfectly fine that there is no revision apart from the current PR that references it.
```
Comment (burmudar) [inline tools/release/README.md:141]: Should we mention that along with the tarbal being removed we also revert the `tools/release/schema_deps.bzl` change - even though it is sort of implied

File: tools/release/README.md [hunk 12]
```diff
@@ -0,0 +1,198 @@
...
+Now, when building the _migrator_, we can't directly use the GCS bucket. Bazel wants a deterministic set of inputs and "all content from the bucket" is not deterministic.
+To satisfy Bazel, we need a fixed input, checksumed, to guarantee that the build is stable. So when we're creating a new release, we simply regenerate that
+tarball based on the schemas we find in the bucket, under `schemas/` and upload it under `dist/`.
+
+Step by step process (work-in-progress):
+
+1. We want to create a new release, which is materialized by a pull-request automatically created by `sg`
+1. `sg release create ...` runs `bazel run //tools/release:generate_schemas_archive -- v5.3.4444`
+1. it fetches all schemas whose version are below 5.3.4444
+1. it copies the current `schema.*.json` files into `v5.3.4444-internal_database.schema.*.json`, to match the convention of the other schemas.
```
Comment (burmudar) [inline tools/release/README.md:129]:
```suggestion
1. it copies the current `schema.*.json` files to `v5.3.4444-internal_database.schema.*.json`, to match the convention of the other schemas.
```

File: tools/release/generate_schemas_archive.sh [hunk 13]
```diff
@@ -0,0 +1,192 @@
...
+echo "--- Uploading tarball to ${bucket}/dist"
+
+# Tarballs are reproducible, but the only reason for which the user would want to overwrite the existing one
+# is to fix a problem. We don't want anyone to run this by accident, so we explicitly ask for confirmation.
+if gsutil -q ls "${bucket}/dist/${output_basename}"; then
+  echo "--- ⚠️  A database schemas tarball already exists for this version"
+  echo "Type OVERWRITE followed by ENTER to confirm you want to overwrite it. Anything else will abort."
+  read -p "Are you sure? " -r
+  echo
+  if [[ "$REPLY" != "OVERWRITE" ]]
```
[the remark this answers is not in the document]
Comment (jhchabran) [reply tools/release/generate_schemas_archive.sh:167]: It works because the `then` is on the following line.

**Review by burmudar — APPROVED** [review]
